Mask detector for a single person in image

In [1]:
from keras.models import Sequential
from keras.layers import Conv2D
from keras.layers import MaxPooling2D
from keras.layers import Flatten
from keras.layers import Dense , Dropout

In [12]:
model = Sequential()

In [13]:

model.add(Conv2D(32 , (7,7), input_shape=(224,224,3) , activation='relu'))


In [14]:
model.add(MaxPooling2D(pool_size=(2,2)))
model.add(Conv2D(32,(7,7), activation = 'relu'))
model.add(MaxPooling2D(pool_size=(2,2)))

In [15]:
model.add(Flatten())

In [16]:

model.add(Dense(units=128 , activation='relu'))
model.add(Dropout(0.6))
model.add(Dense(units=64, activation = 'relu'))
model.add(Dropout(0.3))
model.add(Dense(units=1  , activation='sigmoid'))
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_2 (Conv2D)            (None, 218, 218, 32)      4736      
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 109, 109, 32)      0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 103, 103, 32)      50208     
_________________________________________________________________
max_pooling2d_3 (MaxPooling2 (None, 51, 51, 32)        0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 83232)             0         
_________________________________________________________________
dense_7 (Dense)              (None, 128)               10653824  
_________________________________________________________________
dropout_5 (Dropout)          (None, 128)              

In [17]:
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [18]:
from keras.preprocessing.image import ImageDataGenerator
training_datagen= ImageDataGenerator(rescale=1/255 , shear_range=0.2 , zoom_range=0.2, horizontal_flip=True)

In [19]:
testing_datagen=ImageDataGenerator(rescale=1/255)
training_set=training_datagen.flow_from_directory('/trainset/', 
                                                   target_size=(224,224), batch_size=32 , class_mode= 'binary')
test_set=testing_datagen.flow_from_directory('/testset/',
                                             target_size=(224,224), batch_size=32 , class_mode= 'binary')

Found 1386 images belonging to 2 classes.
Found 6 images belonging to 2 classes.


In [ ]:
model.fit_generator(training_set,steps_per_epoch=1376/32,epochs=25)

Epoch 1/25
43/43 [==============================] - 222s 5s/step - loss: 0.5913 - accuracy: 0.7245
Epoch 2/25
43/43 [==============================] - 222s 5s/step - loss: 0.4144 - accuracy: 0.8479
Epoch 3/25
43/43 [==============================] - 227s 5s/step - loss: 0.2482 - accuracy: 0.9321
Epoch 4/25
43/43 [==============================] - 223s 5s/step - loss: 0.1782 - accuracy: 0.9483
Epoch 5/25
43/43 [==============================] - 222s 5s/step - loss: 0.1370 - accuracy: 0.9638
Epoch 6/25
43/43 [==============================] - 224s 5s/step - loss: 0.1211 - accuracy: 0.9645
Epoch 7/25
43/43 [==============================] - 223s 5s/step - loss: 0.2284 - accuracy: 0.9424
Epoch 8/25
43/43 [==============================] - 227s 5s/step - loss: 0.1456 - accuracy: 0.9542
Epoch 9/25
43/43 [==============================] - 223s 5s/step - loss: 0.0939 - accuracy: 0.9690
Epoch 10/25
43/43 [==============================] - 222s 5s/step - loss: 0.0953 - accuracy: 0.9712
Epoch 11/

In [ ]:
y_pred=model.predict(test_set, batch_size=32)

In [ ]:
acc=model.evaluate(training_set,steps=1386,verbose=0)

In [ ]:
print(acc)

In [ ]:
for i in y_pred:
  if (1-i)>0.6:
    print("Masked")
  else:
    print("not masked")

this model classifies the masked and unmasked faces and has an accuracy of 94%